code/scripting/api/libs/mission.cpp

```diff
@@ -988,6 +988,74 @@ ADE_FUNC(sendPlainMessage,
 	return ADE_RETURN_TRUE;
 }
 
+ADE_FUNC(addMessageToScrollback,
+	l_Mission,
+	"string message, [number|enumeration source=HUD_SOURCE_COMPUTER]",
+	"Adds a string to the message log scrollback without sending it as a message first. Source matches the built-in sources and then "
+	"the team index where (N - num built-in sources) == team index. Currently the built-in sources are 0-7 in this order: Computer, training, hidden, "
+	" important, failed, satisfied, terran command, netplayer.",
+	"boolean",
+	"true if successful, false otherwise")
+{
+	const char* message = nullptr;
+	int source = HUD_SOURCE_COMPUTER;
+	enum_h* esp = nullptr;
+
+	if (lua_isnumber(L, 2)) {
+		if (!ade_get_args(L, "s|i", &message, &source)) {
+			return ADE_RETURN_FALSE;
+		} else {
+			if (source < HUD_SOURCE_TEAM_OFFSET) {
+				LuaError(L, "Source index must be greater than the %i built-in sources.", HUD_SOURCE_TEAM_OFFSET);
+				return ADE_RETURN_FALSE;
+			}
+		}
+	} else {
+		if (!ade_get_args(L, "s|i", &message, l_Enum.GetPtr(&esp))) {
+			return ADE_RETURN_FALSE;
```

Comment:
The function signature of `addMessageToScrollback` must be changed.  Allowing a number is going to cause scripts to break if the hud sources are ever changed.  Instead the argument should be either an enumeration or an IFF type.  The `ade_odata_is_userdata_type` function can be used to check what the type is.

Reply:
Good idea. Didn't know that `is_userdata_type` function existed. Done.